crates/ironrdp-rdpdr/src/lib.rs

```diff
@@ -10,17 +19,29 @@ use tracing::warn;
 #[derive(Debug)]
 pub struct Rdpdr;
 
+impl Default for Rdpdr {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
 impl Rdpdr {
     pub const NAME: ChannelName = ChannelName::from_static(b"rdpdr\0\0\0");
 
     pub fn new() -> Self {
         Self
     }
-}
 
-impl Default for Rdpdr {
-    fn default() -> Self {
-        Self::new()
+    fn handle_server_announce(&mut self, payload: &mut ReadCursor<'_>, output: &mut WriteBuf) -> PduResult<()> {
+        let req = ServerAnnounceRequest::decode(payload)?;
+        trace!("{:?}", req);
+        let res = ClientAnnounceReply::new(28, 0, req.client_id);
+        // TODO:
+        // - common headers?
+        // - chunkification?
+        // - difference between `WriteBuf` and `WriteCursor`?
```

Comment:
The main purpose is to enable the channel implementer to grow the output buffer dynamically, as appropriate. I initially thought about asking the user to advertise the required space prior to the `process` call, but this didn’t sounded like a practical approach because it’s unlikely the user know exactly how much space is required before actually processing the input payload in order to generate a specific response.

That’s where I came up with `WriteCursor`, which some sort of helper for [this kind of code](https://github.com/Devolutions/IronRDP/blob/e10e2f4d91bfd99199457451a810c24ad993c78f/crates/ironrdp-pdu/src/lib.rs#L150-L156):

```rust
    let pdu = /* some PDU structure */;

    let pdu_size = pdu.size();

    if buf.len() < pdu_size {
        buf.resize(pdu_size, 0);
    }

    // ^ this

    let mut cursor = WriteCursor::new(dst);
    encode_cursor(pdu, &mut cursor)?;
```

But it’s going a bit further by keeping track of the already filled region (as opposed to the initialized yet unfilled region of the inner `Vec<u8>`). One purpose is to facilitate buffer re-use: it’s okay to keep around the same `WriteCursor` and repeatedly write into it. Contrary to `Vec<u8>`, bytes will be written at the beginning of the unfilled region (a `Vec<u8>` will write bytes at the beginning of the uninitialized region). This way, it’s easy and safe to retrieve a mutable slice of the unfilled (but already initialized) region. This slice can be used to build a `WriteCursor`. Another purpose of the filled region tracking is to enable multiple items to be written consecutively in the same buffer. This was not possible with the original `encode_buf` function. [The new `encode_buf` function](https://github.com/Devolutions/IronRDP/blob/2eedfe7b8d987d25ec3996738c69d1d0bb3d416d/crates/ironrdp-pdu/src/lib.rs#L158-L165) that is using `WriteCursor` now looks like this:

```rust
pub fn encode_buf<T: PduEncode>(pdu: &T, buf: &mut WriteBuf) -> PduResult<usize> {
    let pdu_size = pdu.size();
    let dst = buf.unfilled_to(pdu_size);
    let mut cursor = WriteCursor::new(dst);
    encode_cursor(pdu, &mut cursor)?;
    let written = cursor.pos();
    buf.advance(written);
    Ok(written)
}
```

A big enough mutable slice into the unfilled region is retrieved by using [the `unfilled_to` helper method](https://github.com/Devolutions/IronRDP/blob/2eedfe7b8d987d25ec3996738c69d1d0bb3d416d/crates/ironrdp-pdu/src/write_buf.rs#L62-L67) (allocating more memory as necessary), a `WriteCursor` wrapping this slice is created and used to actually encode the PDU. We advance the filled region of the `WriteBuf` so that [the `filled` method](https://github.com/Devolutions/IronRDP/blob/2eedfe7b8d987d25ec3996738c69d1d0bb3d416d/crates/ironrdp-pdu/src/write_buf.rs#L50-L52) returns the right region, and subsequent calls do not override it [until `clear` is called](https://github.com/Devolutions/IronRDP/blob/2eedfe7b8d987d25ec3996738c69d1d0bb3d416d/crates/ironrdp-pdu/src/write_buf.rs#L115-L121).

Consider this piece of code:
```rust
fn process(&mut self, initiator_id: u16, channel_id: u16, payload: &[u8], output: &mut WriteBuf) -> PduResult<()> {
    let server_request = ServerRequest: ironrdp_pdu::decode(payload)?;

    match server_request.order {
        ServerOrder::DoThis => {
            // do this
            let response = DoThisResponse { … };
            ironrdp_pdu::encode_buf(response, output)?; // buffer is grown, or not, as appropriate, and `DoThisResponse` is encoded in the "unfilled" region
        }
        ServerOrder::DoThat => {
            // do that
            let response = DoThatResponse { … };
            ironrdp_pdu::encode_buf(response, output)?; // same here
        }
    }

    Ok(())
}
```

[Other methods such as `write_u8`](https://github.com/Devolutions/IronRDP/blob/2eedfe7b8d987d25ec3996738c69d1d0bb3d416d/crates/ironrdp-pdu/src/write_buf.rs#L74-L113) are part of the `WriteBuf` API and, indeed, they are overlapping with the `WriteCursor` API, but it’s mostly for convenience if you ever need to manually write something in an ad-hoc fashion.  Otherwise, using `WriteCursor` is preferable because you get a no-std and no-alloc friendly code right away.

Probably this can be improved upon, but does it makes sense?